subprojects/core/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -0,0 +1,399 @@
+            } finally {
+                withoutWorkerLease(new Runnable() {
+                    @Override
+                    public void run() {
+                        for (DefaultProjectLock projectLock : threadProjectMap.get(threadId)) {
+                            projectLock.lock();
+                        }
+                    }
+                });
+            }
+        } else {
+            runnable.run();
+        }
+    }
+
+    private class DefaultProjectLock implements ProjectLock {
+        private final String projectPath;
+        private ReentrantLock lock = new ReentrantLock();
+
+        DefaultProjectLock(String projectPath) {
+            this.projectPath = projectPath;
+        }
+
+        boolean tryLock() {
+            if (!lock.isHeldByCurrentThread()) {
+                if (lock.tryLock()) {
+                    LOGGER.debug("{}: acquired project lock on {}", Thread.currentThread().getName(), projectPath);
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                return true;
+            }
+        }
+
+        void unlock() {
+            if (lock.isHeldByCurrentThread()) {
+                lock.unlock();
+                LOGGER.debug("{}: released project lock on {}", Thread.currentThread().getName(), projectPath);
+                projectLockBroadcast.onProjectUnlock(projectPath);
+            }
+        }
+
+        void lock() {
+            if (!lock.isHeldByCurrentThread()) {
+                lock.lock();
+                LOGGER.debug("{}: acquired project lock on {}", Thread.currentThread().getName(), projectPath);
+            }
+        }
+
+        @Override
+        public boolean hasProjectLock() {
+            return lock.isHeldByCurrentThread();
+        }
+
+        @Override
+        public boolean isLocked() {
+            return lock.isLocked();
+        }
+
+        @Override
+        public boolean tryWithProjectLock(Runnable runnable) {
+            Long threadId = Thread.currentThread().getId();
+            final DefaultProjectLock thisLock = this;
+            if (!hasProjectLock()) {
+                if (tryLock()) {
+                    try {
+                        threadProjectMap.put(threadId, thisLock);
+                        runnable.run();
+                    } finally {
+                        unlock();
+                        threadProjectMap.remove(threadId, thisLock);
+                    }
+                    return true;
+                }
+            } else {
+                runnable.run();
+                return true;
+            }
+
+            return false;
+        }
+
+        @Override
+        public void withProjectLock(Runnable runnable) {
+            final Long threadId = Thread.currentThread().getId();
+            final DefaultProjectLock thisLock = this;
+            if (!hasProjectLock()) {
+                withoutWorkerLease(new Runnable() {
```

Comment:
I think this demonstrates some of the problems with treating the cpu resource locks and the project resource locks as separate things: eg there's no need to release the cpu resource lock here if the project lock is currently available, and there's no point taking out the cpu resource lock earlier if no project locks are available.

It's all in one service here, at least, so this is pretty much an implementation detail now.